### ISSUE TITLE
The loss doesn't seem to improve after several epochs

### ISSUE THREAD

**Issue by amil-rp-work** [qubvel/segmentation_models#401]
Hey 
Thanks for providing such an easy to use library!

The problem I want to solve is the segmentation of just a single object from the background. My inputs are images of size `HxWx3` and outputs are of size `HxWx1`. The output image has pixel values as 1 denoting the object I want to detect and 0 for all other pixels which denotes background.

[Sample Images](https://imgur.com/a/OdMDyKW)
Please note the mask is completely black since i replaced the pixel of white(255) with 1. It's a grayscale image.

I have referred to the binary segmentation example given in the `README.md` ([Link](https://github.com/qubvel/segmentation_models/blob/master/examples/binary%20segmentation%20(camvid).ipynb)). I have applied similar data augmentation practices just removed the `random_cropping` and `random rotation`.
The problem I am facing is **the loss doesn't decrease and the same with other metrics.**
Here is the code I am using

```
BACKBONE = 'efficientnetb7'
BATCH_SIZE = 8
CLASSES = ['advert']
LR = 0.0001
EPOCHS = 10

preprocess_input = sm.get_preprocessing(BACKBONE)
n_classes = 1 if len(CLASSES) == 1 else (len(CLASSES) + 1)
activation = 'sigmoid' if n_classes == 1 else 'softmax'

model = sm.Unet(BACKBONE, classes=n_classes, activation=activation,
    input_shape=(None, None, 3))

optim = keras.optimizers.Adam(LR)

dice_loss = sm.losses.DiceLoss()
focal_loss = sm.losses.BinaryFocalLoss() if n_classes == 1 else sm.losses.CategoricalFocalLoss()
total_loss = dice_loss + (1 * focal_loss)

metrics = [sm.metrics.IOUScore(threshold=0.5), sm.metrics.FScore(threshold=0.5)]

model.compile(optim, total_loss, metrics)

train_dataset = Dataset(
    x_train_dir, 
    y_train_dir, 
    classes=CLASSES, 
    augmentation=get_training_augmentation(),
    preprocessing=get_preprocessing(preprocess_input),
)

valid_dataset = Dataset(
    x_valid_dir, 
    y_valid_dir, 
    classes=CLASSES, 
    augmentation=get_validation_augmentation(),
    preprocessing=get_preprocessing(preprocess_input),
)

train_dataloader = Dataloder(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
valid_dataloader = Dataloder(valid_dataset, batch_size=1, shuffle=False)

assert train_dataloader[0][0].shape == (BATCH_SIZE, 320, 320, 3)
assert train_dataloader[0][1].shape == (BATCH_SIZE, 320, 320, n_classes)

callbacks = [
    keras.callbacks.ModelCheckpoint('./best_model.h5', save_weights_only=False, save_best_only=True, mode='min'),
    keras.callbacks.ReduceLROnPlateau(),
]

history = model.fit_generator(
    train_dataloader, 
    steps_per_epoch=len(train_dataloader), 
    epochs=EPOCHS, 
    callbacks=callbacks, 
    validation_data=valid_dataloader, 
    validation_steps=len(valid_dataloader)
)
```

**Comment by amil-rp-work** [comment]
@JordanMakesMaps I know you have been continuously helping the community. I would be glad to provide more context about my problem. It would be really great if you could help me out here.
Any help would be appreciated.

**Comment by JordanMakesMaps** [comment]
It seems like you're having a similar issue as a user in the issue #345. Because you're doing a binary classification, you need to include another class to represent all of the other pixels that do not belong to the class that you are interested in. In you're case, you care about 'advert', but the model doesn't know what to do with all of the other pixels. So the issue with your loss and metrics _could_ be because of this; I would make these adjustments to your code and see what the results are:

(please report back, if this works, then I'll look at creating another example script for both binary and multi categorical classification and see if @qubvel will add it.)

Also, you may need to adjust `__getitem__()` in the dataloader class to ensure that your masks are opened, processed and passed to the model during training correctly as shape (Batch, H, W, n_classes  == 2). Again, see #345 and look into `keras.utils.to_categorical()`.

```python
BACKBONE = 'efficientnetb7'
BATCH_SIZE = 8
CLASSES = ['background', 'advert'] # <--- binary == two classes
LR = 0.0001 
EPOCHS = 10

preprocess_input = sm.get_preprocessing(BACKBONE)

n_classes = 2
activation = 'sigmoid'

model = sm.Unet(BACKBONE, classes=n_classes, activation=activation,
    input_shape=(None, None, 3))

optim = keras.optimizers.Adam(LR)

dice_loss = sm.losses.DiceLoss()
focal_loss = sm.losses.BinaryFocalLoss() 
total_loss = dice_loss + (1 * focal_loss)

metrics = [sm.metrics.IOUScore(threshold=0.5), sm.metrics.FScore(threshold=0.5)]

model.compile(optim, total_loss, metrics)

train_dataset = Dataset(
    x_train_dir, 
    y_train_dir, 
    classes=CLASSES, 
    augmentation=get_training_augmentation(),
    preprocessing=get_preprocessing(preprocess_input),
)

valid_dataset = Dataset(
    x_valid_dir, 
    y_valid_dir, 
    classes=CLASSES, 
    augmentation=get_validation_augmentation(),
    preprocessing=get_preprocessing(preprocess_input),
)

train_dataloader = Dataloder(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
valid_dataloader = Dataloder(valid_dataset, batch_size=1, shuffle=False)

assert train_dataloader[0][0].shape == (BATCH_SIZE, 320, 320, 3)
assert train_dataloader[0][1].shape == (BATCH_SIZE, 320, 320, n_classes)

callbacks = [
    keras.callbacks.ModelCheckpoint('./best_model.h5', save_weights_only=False, save_best_only=True, mode='min'),
    keras.callbacks.ReduceLROnPlateau(),
]

history = model.fit_generator(
    train_dataloader, 
    steps_per_epoch=len(train_dataloader), 
    epochs=EPOCHS, 
    callbacks=callbacks, 
    validation_data=valid_dataloader, 
    validation_steps=len(valid_dataloader)
)

```

**Comment by VladislavAD** [comment]
I have successfuly trained Unet for single class segmentation using dice loss a while ago (before v1.0.0 release) without any second class, has something changed? Also I see you define some custom metrics, do you have the same result with something simple like `loss = sm.losses.bce_dice_loss` and  `metrics = [sm.metrics.iou_score]`?

**Comment by amil-rp-work** [comment]
@JordanMakesMaps 
Thanks for the update, I have finally got the loss to work, thanks to you!
Initially, the loss is `~0.6` where the model learns to predict only background pixels and slowly it learns to predict the other class as well and the loss goes to about `~0.25`
Could you suggest which pair of backbone & models are the better-performing ones?
I know no single pair is best hence maybe the ones that seemed to work well with your use cases?
Apart from model & backbone what other methods have you adopted to increase performance?

@VladislavAD 
I haven't had the chance to play with losses, however, the resources I referred from the internet seemed to indicate that Focal loss seems to work well with huge class imbalance, just like in my case. I will give it a try and let you know

**Comment by JordanMakesMaps** [comment]
@amil-rp-work I'd check out the EfficientNet series with `noisy-student` as the pre-trained weights instead of `imagenet` as the encoder, and try the U-Net architecture. The size (b0-b7) depends on the amount of data you have so you'll have to try out different ones. 

As @VladislavAD said, I would try out different losses and compare, I found jaccard loss to do well for my dataset. Other things of course include different types of augmentation, learning rates, optimizers, and batch size. Good luck!

**Comment by amil-rp-work** [comment]
@JordanMakesMaps Thanks a lot for the inputs, will definitely try the pair of model & backbone that you suggested. 
Have you tried freezing and unfreezing the encoder, does it help?
Will report back my findings so it helps the community!

**Comment by JordanMakesMaps** [comment]
With my dataset that I collected of underwater habitats, I found that freezing the encoder entirely for the whole duration of training led to the fastest and best results. Fine-tuning did nothing to help increase classification accuracies; this may be different for your datasets though.

**Comment by amil-rp-work** [comment]
@JordanMakesMaps Thanks for the amazing help !!!!!!!!!
Closing the issue now since it worked and what you suggested would be really helpful

> (please report back, if this works, then I'll look at creating another example script for both binary and multi categorical classification and see if @qubvel will add it.)
